processor/lsmintervalprocessor/internal/merger/value.go

```diff
@@ -0,0 +1,796 @@
+// under the License.
+
+package merger // import "github.com/elastic/opentelemetry-collector-components/processor/lsmintervalprocessor/internal/merger"
+
+import (
+	"errors"
+	"fmt"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/elastic/opentelemetry-collector-components/processor/lsmintervalprocessor/config"
+	"github.com/elastic/opentelemetry-collector-components/processor/lsmintervalprocessor/internal/data"
+	"github.com/elastic/opentelemetry-collector-components/processor/lsmintervalprocessor/internal/identity"
+	"github.com/elastic/opentelemetry-collector-components/processor/lsmintervalprocessor/internal/merger/limits"
+)
+
+const (
+	resourceLimitsEncodingPrefix = "_resource"
+	overflowMetricName           = "_other"
+	overflowMetricDesc           = "Overflow count due to datapoints limit"
+)
+
+// Not safe for concurrent use.
+type Value struct {
+	resourceLimitCfg config.LimitConfig
+	scopeLimitCfg    config.LimitConfig
+	scopeDPLimitCfg  config.LimitConfig
+
+	source pmetric.Metrics
+	// Keeps track of resource metrics overflow
+	resourceLimits *limits.Tracker
+
+	// Lookup tables created from source
+	resLookup      map[identity.Resource]resourceMetrics
+	scopeLookup    map[identity.Scope]scopeMetrics
+	metricLookup   map[identity.Metric]metric
+	numberLookup   map[identity.Stream]numberDataPoint
+	summaryLookup  map[identity.Stream]summaryDataPoint
+	histoLookup    map[identity.Stream]histogramDataPoint
+	expHistoLookup map[identity.Stream]exponentialHistogramDataPoint
+}
+
+type resourceMetrics struct {
+	pmetric.ResourceMetrics
+
+	// Keeps track of scope overflows within each resource metric
+	scopeLimits *limits.Tracker
+}
+
+type scopeMetrics struct {
+	pmetric.ScopeMetrics
+
+	// Keeps track of datapoints limits within each scope metric
+	datapointsLimits *limits.Tracker
+}
+
+type metric = pmetric.Metric
+
+type numberDataPoint = pmetric.NumberDataPoint
+
+type summaryDataPoint = pmetric.SummaryDataPoint
+
+type histogramDataPoint = pmetric.HistogramDataPoint
+
+type exponentialHistogramDataPoint = pmetric.ExponentialHistogramDataPoint
+
+// NewValue creates a new instance of the value with the configured limiters.
+func NewValue(resLimit, scopeLimit, dpLimit config.LimitConfig) Value {
+	return Value{
+		resourceLimitCfg: resLimit,
+		scopeLimitCfg:    scopeLimit,
+		scopeDPLimitCfg:  dpLimit,
+		source:           pmetric.NewMetrics(),
+		resourceLimits:   limits.NewTracker(resLimit.MaxCardinality),
+		resLookup:        make(map[identity.Resource]resourceMetrics),
+		scopeLookup:      make(map[identity.Scope]scopeMetrics),
+		metricLookup:     make(map[identity.Metric]metric),
+		numberLookup:     make(map[identity.Stream]numberDataPoint),
+		summaryLookup:    make(map[identity.Stream]summaryDataPoint),
+		histoLookup:      make(map[identity.Stream]histogramDataPoint),
+		expHistoLookup:   make(map[identity.Stream]exponentialHistogramDataPoint),
```

Comment:
Should we defer instantiating these maps until first use? Particularly for single batches, some of them may never be used.

Reply:
Done!